File: 2015/23_OpeningtheTuringLock/computer.js
/* eslint-disable linebreak-style */
// ======================================================================
// Opening the Turing Lock
//   Advent of Code 2015 Day 23 -- Eric Wastl -- https://adventofcode.com
//
// JavaScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           c o m p u t e r . j s
//
// A solver for the Advent of Code 2015 Day 23 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

const INSTRUCTIONS = {
  hlf: 'r',
  tpl: 'r',
  inc: 'r',
  jmp: 'o',
  jie: 'ro',
  jio: 'ro',
};

// ======================================================================
//                                                               Computer
// ======================================================================

class Computer {
  // Object for Opening the Turing Lock

  constructor(options) {
    // Create a Computer object

    // 1. Set the initial values
    this.text = options.text === undefined ? null : options.text;
    this.part2 = options.part2 === undefined ? false : options.part2;
    this.registers = { a: 0, b: 0, pc: 0 };
    this.instructions = [];

    // 2. Process text (if any)
    if (this.text !== null) {
      this.processText(this.text);
    }
  }

  reset() {
    this.registers = { a: 0, b: 0, pc: 0 };
  }

  processText(text) {
    // 1. Reset the instructions
    this.instructions = [];
    let pc = 0;
    // 2. Loop for all of the lines of text
    text.forEach((line) => {
      // 3. Split out instruction and operands
      let badInst = null;
      let [opcode, op1, op2] = line.split(' ');
      if (opcode) opcode = opcode.trim();
      if (op1) op1 = op1.trim();
      if (op1 && op1[op1.length - 1] === ',') op1 = op1.substring(0, op1.length - 1);
      if (op2) op2 = op2.trim();
      // 4. Check the instruction and operands
      if (!(opcode in INSTRUCTIONS)) {
        badInst = `Unkown opcode ${pc} ${line}`;
      } else {
        const ops = INSTRUCTIONS[opcode];
        if (ops[0] === 'r' && op1 !== 'a' && op1 !== 'b') {
          badInst = `Invalid register operand ${pc} ${line}`;
        } else if (ops[0] === 'o' && Computer.badOffset(op1)) {
          badInst = `Invalid jump offset ${pc} ${line}`;
        } else if (ops.length === 2 && Computer.badOffset(op2)) {
          badInst = `Invalid conditional jump offset ${pc} ${line}`;
        }
        if (badInst === null) {
          if (ops[0] === 'o') op1 = parseInt(op1, 10);
          if (ops.length === 2) op2 = parseInt(op2, 10);
        }
      }
      // 5. Report any problems
      if (badInst !== null) {
        // eslint-disable-next-line no-console
        console.log(badInst);
      } else {
        this.instructions.push({ opcode, op1, op2 });
        pc += 1;
      }
    });
  }

  static badOffset(operand) {
    if (typeof operand !== 'string') return true;
    if (operand.length < 2) return true;
    if (operand[0] !== '+' && operand[0] !== '-') return true;
    if (operand[1] < '0' || operand[1] > '9') return true;
    return false;
  }

  step() {
    // 1. Check the program counter, halt if out of range
    const { a, b, pc } = this.registers;
    if (pc < 0 || pc >= this.instructions.length) return true;
    // 2. Execute the instruction
    const { opcode, op1, op2 } = this.instructions[pc];
    switch (opcode) {
      case 'hlf':
        this.registers[op1] = Math.floor(this.registers[op1] / 2);
        this.registers.pc += 1;
        break;
      case 'tpl':
        this.registers[op1] *= 3;
        this.registers.pc += 1;
        break;
      case 'inc':
        this.registers[op1] += 1;
        this.registers.pc += 1;
        break;
      case 'jmp':
        this.registers.pc += op1;
        break;
      case 'jie':
        this.registers.pc += this.registers[op1] % 2 === 0 ? op2 : 1;
        break;
      case 'jio':
        this.registers.pc += this.registers[op1] === 1 ? op2 : 1;
        break;
      default:
        // eslint-disable-next-line no-console
        console.log(`pc=${pc} invalid opcode ${opcode}`);
        return true;
    }
    return false;
  }

  run(options) {
    // 1. Get the optional arguments
    const verbose = options.verbose === undefined ? false : options.verbose;
    const limit = options.limit === undefined ? 0 : options.limit;
    // eslint-disable-next-line no-console
    // 2. Loop until the program stops;
    let halt = false;
    let knt = 0;
    while (!halt && (limit === 0 || knt < limit)) {
      if (verbose) {
        const { a, b, pc } = this.registers;
        if (pc >= 0 && pc < this.instructions.length) {
          const { opcode, op1, op2 } = this.instructions[pc];
          // eslint-disable-next-line no-console
          console.log(`${pc}: ${opcode} ${op1} ${op2} a = ${a} b = ${b}`);
        }
      }
      halt = this.step();
      knt += 1;
    }
    if (verbose) {
      const { a, b, pc } = this.registers;
      if (halt) {
        // eslint-disable-next-line no-console
        console.log(`${pc}: Halted a = ${a} b = ${b}`);
      } else {
        // eslint-disable-next-line no-console
        console.log(`${pc}: Limit ${limit} a = ${a} b = ${b}`);
      }
    }
  }

  partOne(options) {
    // Returns the solution for part one

    // 0. Function arguments
    // eslint-disable-next-line no-unused-vars
    const verbose = options.verbose === undefined ? false : options.verbose;
    // eslint-disable-next-line no-unused-vars
    const limit = options.limit === undefined ? 0 : options.limit;

    // 1. Run the computer
    this.run({ verbose, limit });
    // 2. Return the solution for part one
    return this.registers.b;
  }

  partTwo(options) {
    // Returns the solution for part two

    // 0. Function arguments
    // eslint-disable-next-line no-unused-vars
    const verbose = options.verbose === undefined ? false : options.verbose;
    // eslint-disable-next-line no-unused-vars
    const limit = options.limit === undefined ? 0 : options.limit;
    this.todo = 'TODO';

    // 1. Run the computer
    this.registers.a = 1;
    this.run({ verbose, limit });
    // 2. Return the solution for part one
    return this.registers.b;
  }
}

module.exports.Computer = Computer;
// ======================================================================
// end                      c o m p u t e r . j s                     end
// ======================================================================

File: 2015/23_OpeningtheTuringLock/computer.py
# ======================================================================
# Opening the Turing Lock
#   Advent of Code 2015 Day 23 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         c o m p u t e r . p y
# ======================================================================
"A solver for the Advent of Code 2015 Day 23 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------

# ======================================================================
#                                                               Computer
# ======================================================================


class Computer(object):   # pylint: disable=R0902, R0205
    "Object for Opening the Turing Lock"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text
        if text is None:
            self.text = []
        self.regs = {'a': 0, 'b': 0}
        self.addr = 0
        self.reset()

    def reset(self):
        "Reset the computer"

        # 1. Restore the initial values
        self.regs = {'a': 0, 'b': 0}
        self.addr = 0

        # 2. Part two has a twist
        if self.part2:
            self.regs['a'] = 1

    def step(self):
        "Execute a single instruction"

        # 1. Get the instruction
        if self.addr < 0 or self.addr >= len(self.text):
            return False
        inst = self.text[self.addr]

        # 2. Break the instruction info parts
        parts = inst.replace(',', '').split()
        opcode = parts[0]
        op1 = parts[1]
        if len(parts) == 3:
            op2 = parts[2]
        else:
            op2 = None
        next_addr = self.addr + 1
        # print(self.addr, inst, opcode, op1, op2, next_addr)

        # 3. Execute the instruction
        if opcode == 'hlf':
            # 3a. hlf r sets register r to half its current value
            self.regs[op1] = self.regs[op1] // 2
        elif opcode == 'tpl':
            # 3b. tpl r sets register r to triple its current value
            self.regs[op1] = self.regs[op1] * 3
        elif opcode == 'inc':
            # 3c. inc r increments register r, adding 1 to it
            self.regs[op1] = self.regs[op1] + 1
        elif opcode == 'jmp':
            # 3d. jmp offset is a relative jump
            next_addr = self.addr + int(op1)
        elif opcode == 'jie':
            # 3e. jie r, offset is like jmp, but if reg r is even
            if self.regs[op1] % 2 == 0:
                next_addr = self.addr + int(op2)
        elif opcode == 'jio':
            # 3f. jio r, offset is like jmp, but if reg r is one
            if self.regs[op1] == 1:
                next_addr = self.addr + int(op2)
        else:
            print("Bad inst", self.addr, inst, opcode, op1, op2, next_addr)
            return False

        # 4. Set the program counter to its next value
        self.addr = next_addr

        # 5. Return success
        return True

    def run(self):
        "Run the program"

        # 1. Loop with the program counter is good
        while self.step():
            pass

        # 2. Return the b register
        return self.regs['b']

    def part_one(self, verbose=False, limit=0):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.run()

    def part_two(self, verbose=False, limit=0):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.run()


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                      c o m p u t e r . p y                     end
# ======================================================================
